### PR TITLE
FIX: fixing training hanging

### PR DESCRIPTION
**Pull request by razmikmelikbekyan**
*Issue #329 , if available:*

*Description of changes:*
- Implementation of `CustomProcess` class, which has property `exception`.
- Using `CustomProcess` class for individual training process.
- Checking individual training fails due to some Exception using `exception` property of `CustomProcess`.
- Before fetching from local reporter queue checking if it has some results, if not continue waiting for results or exception in `while self.process.is_alive()`. 
- If process is not fails,  then reported results are transferred to `dist_reporter` and so on.
- if process fails, terminating `local_reporter` and transferring exception traceback to `dist_reporter`.

The change is implemented only for `FIFO` scheduler, it will be approved, then we can implement also for other schedulers. 

Code sample # 1:
```
import autogluon as ag
import pandas as pd
import sklearn.datasets
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import r2_score, mean_absolute_error

initial_data = sklearn.datasets.load_boston(return_X_y=False)
data = pd.DataFrame(initial_data['data'], columns=initial_data['feature_names'])
data['target'] = initial_data['target']


def train_rf(args, reporter):
    train_data = args.train_data
    target_variable = args.target_variable

    rf = RandomForestRegressor(
        n_estimators=args.n_estimators,
        max_depth=args.max_depth,
        min_samples_leaf=args.min_samples_leaf,
        min_samples_split=args.min_samples_split,
        min_weight_fraction_leaf=args.min_weight_fraction_leaf,
        random_state=34
    )

    rf.fit(train_data.drop(columns=target_variable), train_data[target_variable])
    pred = rf.predict(train_data.drop(columns=target_variable))
    mae = abs(mean_absolute_error(train_data[target_variable], pred))
    r2 = r2_score(train_data[target_variable], pred)
    reporter(mae=mae, r2=r2)


@ag.args(
    n_estimators=-10,  # this will raise an exception, since it must be >= 1
    max_depth=ag.space.Int(1, 10),
    min_samples_leaf=ag.space.Int(1, 5),
    min_samples_split=ag.space.Int(2, 5),
    min_weight_fraction_leaf=ag.space.Real(0.0, 1e-1),
    train_data=data,
    target_variable="target"
)
def ag_train_rf(args, reporter):
    return train_rf(args, reporter)


myscheduler = ag.scheduler.FIFOScheduler(
    ag_train_rf,
    resource={'num_cpus': 8, 'num_gpus': 0},
    num_trials=5,
    reward_attr='r2',
    searcher='skopt',  # currently supports "random", "skopt", "gird",
    max_reward=0,
)
print(myscheduler)
myscheduler.run()
myscheduler.join_jobs()
```

Code sample # 2:
```
import autogluon as ag
from autogluon import ImageClassification as task

train_data = task.Dataset(name='fashionmnist')

classifier = task.fit(
    train_data,
    num_trials=1,
    net=ag.Categorical('hello'),  # this will raise an exception
    epochs=5,
    ngpus_per_trial=1,
    verbose=False,
    auto_search=False,
)
```

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by szha** [comment]
Job PR-345-1 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-345/1/index.html

File: autogluon/scheduler/reporter.py [hunk 1]
```diff
@@ -174,15 +174,29 @@ def __init__(self, process, local_reporter, dist_reporter):
 
     def run(self):
         while self.process.is_alive():
+
+            # breaking communication if process raises exception
+            if self.process.exception is not None:
+                error, traceback = self.process.exception
+                self.local_reporter.terminate()
+                self.dist_reporter(done=True, traceback=traceback)
```
Comment (mseeger) [inline autogluon/scheduler/reporter.py]: Not sure if done=True is OK here. In the end, the process aborted. done=True I think implies it stopped normally and published a metric, no?

File: autogluon/scheduler/reporter.py [hunk 2]
```diff
@@ -174,15 +174,29 @@ def __init__(self, process, local_reporter, dist_reporter):
 
     def run(self):
         while self.process.is_alive():
+
+            # breaking communication if process raises exception
+            if self.process.exception is not None:
+                error, traceback = self.process.exception
```
Comment (mseeger) [inline autogluon/scheduler/reporter.py]: error is not being used

File: autogluon/scheduler/reporter.py (hunk 1 above)
Comment (razmikmelikbekyan) [reply autogluon/scheduler/reporter.py]: In general I agree with you that `done=True` is a bit confusing. We can remove it. It should work without it also.

File: autogluon/scheduler/reporter.py (hunk 2 above)
Comment (razmikmelikbekyan) [reply autogluon/scheduler/reporter.py]: yep, since its traceback already contains all the info. I can replace it with `_`, but maybe you have other ideas how we can use it here. I was thinking maybe we can raise this error somewhere.

Comment (mseeger) [reply autogluon/scheduler/reporter.py]: If the error tells me the type of exception, it should also be transmitted. It cannot hurt, even though most generic schedulers will not be able to use it.

Comment (razmikmelikbekyan) [reply autogluon/scheduler/reporter.py]: so where do you think is most appropriate to use it?

**Comment by Innixma** [comment]
Would this enable Tabular to not have to wrap try/except in its HPO code?

**Comment by razmikmelikbekyan** [comment]
> Would this enable Tabular to not have to wrap try/except in its HPO code?

Yes, I checked it with some examples from that fix.

File: autogluon/scheduler/reporter.py (hunk 2 above)
Comment (mseeger) [reply autogluon/scheduler/reporter.py]: Well, normally a scheduler would just note that the evaluation failed, and it would react to any such failure in the same way. Only highly specific schedulers could do more than that.
But certainly, we should log everything we can about what happened, it may be useful.

File: autogluon/scheduler/fifo.py [hunk 3]
```diff
@@ -257,6 +257,10 @@ def _run_reporter(self, task, task_job, reporter, searcher):
         last_result = None
         while not task_job.done():
             reported_result = reporter.fetch()
+            if 'traceback' in reported_result:
```
Comment (mseeger) [inline autogluon/scheduler/fifo.py]: I can confirm that the reaction to an exception being reported (the signal is whether 'traceback' is present) is OK. What happens is that the _run_reporter while loop is left. The current reported_result is not used as a datapoint for the model (this is correct -- it does not have any metric value). What *is* used is the last recently fetched reported_result which did not fail. If the process reports an exception before ever obtaining a metric value, then nothing is reported to the model.

For most searchers using FIFO scheduler, this should be a good reaction. We want to make sure that a config whose evaluation has failed, is not attempted again. At least for RandomSearcher, this currently works, because a config is never attempted twice in the first place.

**Comment by mseeger** [comment]
Given this goes in, I will later add some other method to BaseSearcher, which the scheduler calls if an evaluation failed. While the current searchers maybe don't need that, more advanced searchers would then be able to react. For example, they could explicitly blacklist the config, or even better, they could model the success/failure probability, and then this would be data.
